docs/src/main/sphinx/develop/supporting-merge.rst

```diff
@@ -0,0 +1,390 @@
+        customer VARCHAR,
+        purchases DECIMAL,
+        address VARCHAR);
+    INSERT INTO source_table (customer, purchases, address) VALUES ...;
+
+Here is a possible ``MERGE`` operation, from ``source_table`` to
+``target_table``::
+
+    MERGE INTO target_table t USING source_table s
+        ON (t.customer = s.customer)
+        WHEN MATCHED AND s.address = 'Berkeley' THEN
+            DELETE
+        WHEN MATCHED AND s.customer = 'Joe Shmoe' THEN
+            UPDATE SET purchases = purchases + 100.0
+        WHEN MATCHED THEN
+            UPDATE
+                SET purchases = s.purchases + t.purchases, address = s.address
+        WHEN NOT MATCHED THEN
+            INSERT (customer, purchases, address)
+                VALUES (s.customer, s.purchases, s.address);
+
+SQL ``MERGE`` tries to match each ``WHEN`` clause in source order.  When
+a match is found, the corresponding ``DELETE``, ``INSERT`` or ``UPDATE``
+is executed and subsequent ``WHEN`` clauses are ignored.
+
+SQL ``MERGE`` supports two operations on the target table and source
+when a row from the source table or query matches a row in the target table:
+
+* ``UPDATE``, in which the columns in the target row are updated.
+* ``DELETE``, in which the target row is deleted.
+
+In the ``NOT MATCHED`` case, SQL ``MERGE`` supports only ``INSERT``
+operations.  The values inserted are arbitrary but usually come from
+the unmatched row of the source table or query.
+
+``RowChangeParadigm``
+---------------------
+
+Different connectors have different ways of representing row updates,
+imposed by the underlying storage systems.  The  Trino engine classifies
+these different paradigms as elements of the ``RowChangeParadigm``
+enumeration, returned by ``ConnectorMetadata`` method
+``getRowChangeParadigm(...)``.
+
+The ``RowChangeParadigm`` enumeration values are:
+
+* ``CHANGE_ONLY_UPDATED_COLUMNS``, intended for connectors like
+  ``trino-jdbc`` and ``trino-kudu`` that can update individual columns of
+  rows identified by a ``rowId``.  The corresponding merge processor class is
+  ``ChangeOnlyUpdatedColumnsMergeProcessor``.
+* ``DELETE_ROW_AND_INSERT_ROW``, intended for connectors like ``trino-hive``
+  and ``trino-iceberg`` that represent a row change as a row deletion paired
+  with a row insertion.  The corresponding merge processor class is
+  ``DeleteAndInsertMergeProcessor``.
+
+Overview of ``MERGE`` processing
+--------------------------------
+
+A ``MERGE`` statement is processed by creating a ``RIGHT JOIN`` between the
+target table and the source, on the ``MERGE`` criteria.  The source may be
+a table or an arbitrary query.  For each row in the source table or query,
+``MERGE`` produces a ``ROW`` object containing:
+
+* the column values from the ``UPDATE`` or ``INSERT`` cases, or ``NULL``
+  for the ``DELETE`` cases
+* an integer providing the case number of the case that matched the source
+  row, or -1 meaning that no case matched the source row
```

Comment:
The docs were confused.  It's not a "case number", it's the boolean "present" column, true if the source row matched a target row and false otherwise.

I've fixed the documentation.